### DISCUSSION TITLE
Is there possible to load the html tag in textframe

### DISCUSSION THREAD

**Discussion by DarwinRajagopalan** [ShapeCrawler/ShapeCrawler#543] [Q&A]
**_Example_** :-

`String question = "<p>Hai, <strong> how are you ? </strong></p>";`

**_output :_** Hai, **how are you ?**   // like this output  need to set in ppt text frame 

i dont like to use sc font property  for set the font styles instead of that i would like to  use the innerhtml tag  

because lot of condition have to write  based on html tag and then assign into scfont  .

**Note :-** am using **Rich textbox**, that textbox string values are directly load into the ppt . that's why am asking 
@huchim @adamshakhabov

**Comment by huchim** [comment]
Not really... or maybe not in the scope of this library...

`p, strong` are tokens in an AST and you would need an ast parser and a "renderer"... but `TextFrame` can only contains texts, maybe `p, span, strong, em, \n` could be possible...

**Example:**

```cs
var html = "<p>hai, <strong> how are you?</strong></p>";
var ast = parser(html);
var renderer = new Renderer(shape.TextFrame!);

foreach (var token in ast) {
  renderer.Render(token);
}

class Renderer {
  void Render(T token) {
    switch (token.GetType()) {
      case "P": this.RenderParagraph(token)
      ...
    }
  }

  void RenderParagraph(T token) { 
     _activeP = _textFrame.Paragraphs.Add();
  }

  var RenderInline(InlineToken token) {
   _activeP.Portions[0].Text = token;
  }
  void RenderStrong(StrongToken token) { ... }
  void RenderContainer(ContainerToken token) { 
    // recursive
   foreach (var t in token) {
      Render(t);
   }
  }
}
```

As you can see, you will need defines types for token and specific codes for that tokens.. and after that, you will need to parse `style` attributes (inline CSS)...


This is my only opinion :face_with_spiral_eyes:

**Reply by DarwinRajagopalan** [reply]
Thanks lot @huchim  , will work on this and then will let u know my status once I complete

**Comment by DarwinRajagopalan** [comment]
`using HtmlAgilityPack;
private void InnerHtmlBind(string value, ITextFrame textFrame, CustomFont customFont)
        {
            var document = new HtmlDocument();
            document.LoadHtml(value);
            IterationNodes(document.DocumentNode.ChildNodes, textFrame, customFont);

        }
        private void IterationNodes(HtmlNodeCollection htmlNodes, ITextFrame textFrame, CustomFont customFont)
        {

            foreach (var child in htmlNodes)
            {
                if (!string.IsNullOrEmpty(textFrame!.Paragraphs.Last().Text))
                    textFrame!.Paragraphs.Add();
                switch (child.Name)
                {
                    case "ul":
                        RenderBullet(textFrame, child, customFont);
                        break;
                    case "li":
                        IterationNodes(child.ChildNodes, textFrame, customFont);
                        break;
                    case "p":
                       IterationNodes(child.ChildNodes, textFrame, customFont);
                        break;
                    case "strong":
                        RenderStrong(textFrame, child, customFont);
                        break;
                    case "span":
                        RenderSpan(textFrame, child, customFont);
                        break;
                    case "em":
                        RenderEm(textFrame, child, customFont);
                        break;
                    default:
                        SetPortion(textFrame, child.InnerText, customFont);
                        
                        break;
                }

            }
        }
    private void SetPortion(ITextFrame textFrame, string text, CustomFont font)
        {
            if (string.IsNullOrEmpty(text))
                return;
            textFrame!.Paragraphs.Last().AddPortion(text);
            var portion = textFrame!.Paragraphs.Last().Portions.Last();
            portion.Font.Size = font.Size;
            portion.Font.IsBold = font.IsBold;
            portion.Font.IsItalic = font.IsItalic;
            portion.Font.ColorFormat.SetColorByHex(font.HexColor);

        }
  private void RenderBullet(ITextFrame textFrame, HtmlNode htmlNode, CustomFont customFont)
        {
            //if(htmlNode.ParentNode.Name != "li")
            // textFrame!.Paragraphs.Add();
            

            var bullet = textFrame!.Paragraphs.Last().Bullet;
            bullet.Type = SCBulletType.Character;
            bullet.Character = "•";
            bullet.Size = 100;
            IterationNodes(htmlNode.ChildNodes, textFrame, customFont);
            textFrame!.Paragraphs.Add().Bullet.Type = SCBulletType.None;


        }
        private void RenderSpan(ITextFrame textFrame, HtmlNode htmlNode, CustomFont customFont)
        {
            customFont.HexColor = customFont.HexColor == "#000000" ? SetColorRGBtoHex(htmlNode) : customFont.HexColor;
            IterationNodes(htmlNode.ChildNodes, textFrame, customFont);
            customFont.IsBold = false;
            customFont.IsItalic = false;
            customFont.HexColor = "#000000";

        }
        private void RenderStrong(ITextFrame textFrame, HtmlNode htmlNode, CustomFont customFont)
        {
            customFont.IsBold = true;
            customFont.HexColor = customFont.HexColor == "#000000" ? SetColorRGBtoHex(htmlNode) : customFont.HexColor;
            IterationNodes(htmlNode.ChildNodes, textFrame, customFont);
            customFont.IsBold = false;
            customFont.IsItalic = false;
            customFont.HexColor = "#000000";
        }
        private void RenderEm(ITextFrame textFrame, HtmlNode htmlNode, CustomFont customFont)
        {
            customFont.IsItalic = true;
            customFont.HexColor = customFont.HexColor == "#000000" ? SetColorRGBtoHex(htmlNode) : customFont.HexColor;
            IterationNodes(htmlNode.ChildNodes, textFrame, customFont);
            customFont.IsBold = false;
            customFont.IsItalic = false;
            customFont.HexColor = "#000000";
        }
        private string SetColorRGBtoHex(HtmlNode htmlNode)
        {
            var attrFontColor = htmlNode.Attributes.FirstOrDefault(x => x.Value.Contains("rgb"));
            if (attrFontColor != null)
            {
                return Common.GetColorRGBtoHex(attrFontColor.Value);
            }
            return "#000000";
        }
`

**Reply by DarwinRajagopalan** [reply]
@huchim  I done this and thanks for your ideas..
